.Net SDK Validation

MattFromRVA opened the pull request:
Fixes #2199

timcassell commented on src/BenchmarkDotNet/Toolchains/DotNetSdkProvider.cs:
```diff
@@ -0,0 +1,75 @@
+            }
+        }
+
+        private IEnumerable<string> GetInstalledFrameworkSdks()
```
Perhaps this should be pulled out to a separate `DotNetFrameworkSdkProvider`. It should only be necessary for `CsProjClassicNetToolchain`.

timcassell reviewed (changes requested):
`MonoAotLLVMToolChain` and `CsProjClassicNetToolchain` also need the validation.

timcassell commented on src/BenchmarkDotNet/Toolchains/DotNetSdkProvider.cs:
```diff
@@ -0,0 +1,75 @@
+
+            using (var process = Process.Start(startInfo))
+            {
+                if (process == null) throw new InvalidOperationException("Failed to start the dotnet process.");
```
I think this should return empty instead of throw. You can check `HostEnvironmentInfo.GetCurrent().IsDotNetCliInstalled()` before running the process.

MattFromRVA replied:
Ok I just committed some changes with separated .Net and .Net Framework providers. Would you want me to cache both results to `HostEnvironmentInfo`?

timcassell replied:
I guess it doesn't make sense to cache the core versions in `HostEnvironmentInfo` if a custom cli can be used. I'm not sure how expensive retrieving the Framework sdks is, but it would probably be good to cache the results in case of multiple jobs using the same toolchain. The cache could live in the `DotNetSdkVersionValidator` rather than `HostEnvironmentInfo` I guess (and you could cache core sdks per custom cli if you think it's worth it).

timcassell commented on src/BenchmarkDotNet/Validators/DotNetSdkVersionValidator.cs:
```diff
@@ -0,0 +1,102 @@
+
+namespace BenchmarkDotNet.Validators
+{
+    public class DotNetSdkVersionValidator : IValidator
```
I'm not sure if this needs to be a public `IValidator`. I think it could just be a static utility class that the toolchains call. If you want public validators, do separate ones for Core and for Framework. But I don't think it's necessary since the toolchains automatically do the validation.

timcassell commented on src/BenchmarkDotNet/Validators/DotNetSdkVersionValidator.cs:
```diff
@@ -0,0 +1,102 @@
+            string requiredSdkVersion = GetSdkVersionFromMoniker(runtimeMoniker);
+
+            // If the required SDK version is for .NET Framework
+            if (requiredSdkVersion.StartsWith("4"))
```
Just make a `public static IEnumerable<ValidationError> ValidateFrameworkSdks()` and `public static IEnumerable<ValidationError> ValidateCoreSdks(string? customDotNetCliPath)` that the toolchains can call.

MattFromRVA replied:
ok great idea but concerning the issue in #2199, we want to verify that the SDK specified with `RuntimeMoniker` is installed. So are you ok if I add `BenchmarkCase benchmarkCase` as an additional parameter in these static methods? This will allow me access to `benchmarkCase.Job.Environment.Runtime?.RuntimeMoniker`

timcassell replied:
Yes of course that's fine.

timcassell commented on src/BenchmarkDotNet/Toolchains/IDotNetFrameworkSdkProvider.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace BenchmarkDotNet.Toolchains
+{
+    public interface IDotNetFrameworkSdkProvider
```
Creating another interface for this seems pointless. I don't think the interface is necessary at all (see comments in `DotNetSdkVersionValidator`).

timcassell commented:
Another thing you could do is integrate the `Toolchain.IsCliPathInvalid` into this validator.

timcassell commented on src/BenchmarkDotNet/Validators/DotNetSdkVersionValidator.cs:
```diff
@@ -0,0 +1,326 @@
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Interoperability", "CA1416:Validate platform compatibility", Justification = "This code is protected with a runtime OS platform check")]
+        private static void Get1To45VersionFromRegistry(List<string> versions)
```
I don't think this is necessary since we only support net461+.

timcassell commented on src/BenchmarkDotNet/Validators/DotNetSdkVersionValidator.cs:
```diff
@@ -0,0 +1,326 @@
+
+            if (!installedSdks.Any(sdk => sdk.StartsWith(requiredSdkVersion)))
+            {
+                return new ValidationError[]
+                {
+                    new ValidationError(true, $"The required .NET Core SDK version {requiredSdkVersion} for runtime moniker {runtimeMoniker} is not installed.", benchmark)
+                };
```
I think you should return a different `ValidationError` if the `customDotNetCliPath` is invalid (like `Toolchain.IsCliPathInvalid`, then that function can be removed from `Toolchain`).

timcassell commented on src/BenchmarkDotNet/Validators/DotNetSdkVersionValidator.cs:
```diff
@@ -0,0 +1,326 @@
+
+namespace BenchmarkDotNet.Validators
+{
+    public static class DotNetSdkVersionValidator
```
```suggestion
    public static class DotNetSdkValidator
```

timcassell commented on src/BenchmarkDotNet/Validators/DotNetSdkVersionValidator.cs:
```diff
@@ -0,0 +1,326 @@
+            return Enumerable.Empty<ValidationError>();
+        }
+
+        public static IEnumerable<ValidationError> ValidateFrameworkSdks(BenchmarkCase benchmark)
```
This could also use the `IsCliPathInvalid` check since the `CsProjClassicNetToolchain` uses it. Or it could be a separate helper function moved to this class and the toolchain can call both.

timcassell commented on src/BenchmarkDotNet/Validators/DotNetSdkVersionValidator.cs:
```diff
@@ -0,0 +1,326 @@
+
+        private static string GetSdkVersionFromMoniker(RuntimeMoniker runtimeMoniker)
+        {
+            switch (runtimeMoniker)
+            {
+                case RuntimeMoniker.Net461:
+                    return "4.6.1";
+
+                case RuntimeMoniker.Net462:
+                    return "4.6.2";
+
```
Nit: this could be more compact without reducing legibility.
```cs
            return runtimeMoniker switch
            {
                RuntimeMoniker.Net461 => "4.6.1",
                RuntimeMoniker.Net462 => "4.6.2",

```

timcassell reviewed (commented):
Almost there, few more comments.

timcassell commented on src/BenchmarkDotNet/Validators/DotNetSdkVersionValidator.cs:
```diff
@@ -0,0 +1,326 @@
+{
+    public static class DotNetSdkVersionValidator
+    {
+        private static readonly Lazy<List<string>> cachedFrameworkSdks =
+        new Lazy<List<string>>(() => GetInstalledFrameworkSdks().ToList(), true);
```
Just return the expected type from the function.

```suggestion
        private static readonly Lazy<List<string>> cachedFrameworkSdks = new Lazy<List<string>>(GetInstalledFrameworkSdks, true);
```

timcassell commented on src/BenchmarkDotNet/Toolchains/CsProj/CsProjCoreToolchain.cs:
```diff
@@ -49,7 +49,7 @@ public override IEnumerable<ValidationError> Validate(BenchmarkCase benchmarkCas
             }
 
-            if (IsCliPathInvalid(CustomDotNetCliPath, benchmarkCase, out var invalidCliError))
+            if (DotNetSdkValidator.IsCliPathInvalid(CustomDotNetCliPath, benchmarkCase, out var invalidCliError))
```
This can be removed since you added it to `ValidateCoreSdks`.

timcassell commented on src/BenchmarkDotNet/Toolchains/MonoWasm/WasmToolchain.cs:
```diff
@@ -32,10 +32,15 @@ public override IEnumerable<ValidationError> Validate(BenchmarkCase benchmarkCas
                     benchmarkCase);
             }
-            else if (IsCliPathInvalid(CustomDotNetCliPath, benchmarkCase, out var invalidCliError))
+            else if (DotNetSdkValidator.IsCliPathInvalid(CustomDotNetCliPath, benchmarkCase, out var invalidCliError))
```
Same

timcassell commented on src/BenchmarkDotNet/Validators/DotNetSdkVersionValidator.cs:
```diff
@@ -0,0 +1,234 @@
+
+        private static bool HasRuntimeMoniker(BenchmarkCase benchmark)
+        {
+            return !(benchmark == null || benchmark.Job == null || benchmark.Job.Environment?.Runtime?.RuntimeMoniker == null);
```
```suggestion
            return benchmark?.Job?.Environment?.Runtime?.RuntimeMoniker != null;
```

timcassell replied:
May as well make this a `TryGetRuntimeMoniker` method.

timcassell commented on src/BenchmarkDotNet/Toolchains/CsProj/CsProjClassicNetToolchain.cs:
```diff
@@ -50,10 +50,15 @@ public override IEnumerable<ValidationError> Validate(BenchmarkCase benchmarkCas
                 yield return invalidCliError;
             }
+
+            foreach (var validationError in DotNetSdkValidator.ValidateFrameworkSdks(benchmarkCase))
```
This is not protected by the `IsWindows` check.

timcassell commented:
Tests are failing

```
Failed CanDisassembleAllMethodCalls(jit: LegacyJit, platform: X86, runtime: .NET Framework 4.6.2) [13 ms]
  Error Message:
   The "Summary" should have NOT "HasCriticalValidationErrors"
  Stack Trace:
     at BenchmarkDotNet.IntegrationTests.BenchmarkTestExecutor.CanExecute(Type type, IConfig config, Boolean fullValidation) in D:\a\BenchmarkDotNet\BenchmarkDotNet\tests\BenchmarkDotNet.IntegrationTests\BenchmarkTestExecutor.cs:line 70
   at BenchmarkDotNet.IntegrationTests.DisassemblyDiagnoserTests.CanDisassembleAllMethodCalls(Jit jit, Platform platform, Runtime runtime) in D:\a\BenchmarkDotNet\BenchmarkDotNet\tests\BenchmarkDotNet.IntegrationTests\DisassemblyDiagnoserTests.cs:line 101
  Standard Output Messages:
 // Validating benchmarks:
 // Benchmark WithCalls.Benchmark: Dry(Jit=LegacyJit, Platform=X86, Runtime=.NET Framework 4.6.2, IterationCount=1, LaunchCount=1, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=1) [someArgument=2147483647]
 //    * The required .NET Framework SDK version 4.6.2 for runtime moniker Net462 is not installed.
```

Maybe some faulty logic with the Framework sdk validation (maybe newer Framework sdks are backwards-compatible?).

timcassell commented on src/BenchmarkDotNet/Validators/DotNetSdkVersionValidator.cs:
```diff
@@ -0,0 +1,234 @@
+{
+    internal static class DotNetSdkValidator
+    {
+        private static readonly Lazy<IEnumerable<string>> cachedFrameworkSdks = new Lazy<IEnumerable<string>>(GetInstalledFrameworkSdks, true);
```
```suggestion
        private static readonly Lazy<List<string>> cachedFrameworkSdks = new Lazy<List<string>>(GetInstalledFrameworkSdks, true);
```

timcassell commented on src/BenchmarkDotNet/Validators/DotNetSdkVersionValidator.cs:
```diff
@@ -0,0 +1,234 @@
+            }
+        }
+
+        public static IEnumerable<string> GetInstalledFrameworkSdks()
```
```suggestion
        public static List<string> GetInstalledFrameworkSdks()
```

MattFromRVA commented:
> Tests are failing
> 
> ```
> Failed CanDisassembleAllMethodCalls(jit: LegacyJit, platform: X86, runtime: .NET Framework 4.6.2) [13 ms]
>   Error Message:
>    The "Summary" should have NOT "HasCriticalValidationErrors"
>   Stack Trace:
>      at BenchmarkDotNet.IntegrationTests.BenchmarkTestExecutor.CanExecute(Type type, IConfig config, Boolean fullValidation) in D:\a\BenchmarkDotNet\BenchmarkDotNet\tests\BenchmarkDotNet.IntegrationTests\BenchmarkTestExecutor.cs:line 70
>    at BenchmarkDotNet.IntegrationTests.DisassemblyDiagnoserTests.CanDisassembleAllMethodCalls(Jit jit, Platform platform, Runtime runtime) in D:\a\BenchmarkDotNet\BenchmarkDotNet\tests\BenchmarkDotNet.IntegrationTests\DisassemblyDiagnoserTests.cs:line 101
>   Standard Output Messages:
>  // Validating benchmarks:
>  // Benchmark WithCalls.Benchmark: Dry(Jit=LegacyJit, Platform=X86, Runtime=.NET Framework 4.6.2, IterationCount=1, LaunchCount=1, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=1) [someArgument=2147483647]
>  //    * The required .NET Framework SDK version 4.6.2 for runtime moniker Net462 is not installed.
> ```
> 
> Maybe some faulty logic with the Framework sdk validation (maybe newer Framework sdks are backwards-compatible?).

You are correct. Framework sdks are backwards-compatible and it only shows the latest in the registry. I have updated my code to reflect that. 

> Tests are failing
> 
> ```
> Failed CanDisassembleAllMethodCalls(jit: LegacyJit, platform: X86, runtime: .NET Framework 4.6.2) [13 ms]
>   Error Message:
>    The "Summary" should have NOT "HasCriticalValidationErrors"
>   Stack Trace:
>      at BenchmarkDotNet.IntegrationTests.BenchmarkTestExecutor.CanExecute(Type type, IConfig config, Boolean fullValidation) in D:\a\BenchmarkDotNet\BenchmarkDotNet\tests\BenchmarkDotNet.IntegrationTests\BenchmarkTestExecutor.cs:line 70
>    at BenchmarkDotNet.IntegrationTests.DisassemblyDiagnoserTests.CanDisassembleAllMethodCalls(Jit jit, Platform platform, Runtime runtime) in D:\a\BenchmarkDotNet\BenchmarkDotNet\tests\BenchmarkDotNet.IntegrationTests\DisassemblyDiagnoserTests.cs:line 101
>   Standard Output Messages:
>  // Validating benchmarks:
>  // Benchmark WithCalls.Benchmark: Dry(Jit=LegacyJit, Platform=X86, Runtime=.NET Framework 4.6.2, IterationCount=1, LaunchCount=1, RunStrategy=ColdStart, UnrollFactor=1, WarmupCount=1) [someArgument=2147483647]
>  //    * The required .NET Framework SDK version 4.6.2 for runtime moniker Net462 is not installed.
> ```
> 
> Maybe some faulty logic with the Framework sdk validation (maybe newer Framework sdks are backwards-compatible?).

You are correct. Framework sdks are backwards-compatible and it only shows the latest in the registry. I have updated my code to reflect that.

timcassell reviewed (approved):
Thanks a lot @MattFromRVA!

caaavik-msft commented:
This change causes wasm runs to fail (and potentially others I haven't tested) because not all runtime monikers are accounted for in `GetSdkVersionFromMoniker`. I have a PR that adds an integration test for wasm runs which reproduced this and I have included a fix for it by updating `GetSdkVersionFromMoniker` there. I think there are some other monikers missing such as the MonoAotLLVM?

timcassell commented:
> This change causes wasm runs to fail (and potentially others I haven't tested) because not all runtime monikers are accounted for in `GetSdkVersionFromMoniker`. I have a [PR that adds an integration test](https://github.com/dotnet/BenchmarkDotNet/pull/2532) for wasm runs which reproduced this and I have included a fix for it by updating `GetSdkVersionFromMoniker` there. I think there are some other monikers missing such as the MonoAotLLVM?

Oops. This is why we need those integration tests. If someone sends another PR with that fix, I can approve and merge (I can't merge my own PRs).

caaavik-msft commented:
PR up now: https://github.com/dotnet/BenchmarkDotNet/pull/2538
